python-packages/aws-event-stream/aws_event_stream/events.py

```diff
@@ -0,0 +1,514 @@
+        if value:
+            self._buffer.write(b"\x00")
+        else:
+            self._buffer.write(b"\x01")
+
+    def encode_byte(self, value: int) -> None:
+        self._buffer.write(b"\x02")
+        self._buffer.write(pack("!b", value))
+
+    def encode_short(self, value: int) -> None:
+        self._buffer.write(b"\x03")
+        self._buffer.write(pack("!h", value))
+
+    def encode_integer(self, value: int) -> None:
+        self._buffer.write(b"\x04")
+        self._buffer.write(pack("!i", value))
+
+    def encode_long(self, value: int) -> None:
+        self._buffer.write(b"\x05")
+        self._buffer.write(pack("!q", value))
+
+    def encode_blob(self, value: bytes) -> None:
+        # Byte arrays are prefaced with a 16bit length, but are restricted
+        # to a max length of 2**15 - 1
+        if len(value) > MAX_HEADER_VALUE_BYTE_LENGTH:
+            raise InvalidHeaderValueLength(len(value))
+
+        self._buffer.write(b"\06")
+        self._buffer.write(pack("!H", len(value)))
+        self._buffer.write(value)
+
+    def encode_string(self, value: str) -> None:
+        utf8_string = value.encode("utf-8")
+
+        # Strings are prefaced with a 16bit length, but are restricted
+        # to a max length of 2**15 - 1
+        if len(utf8_string) > MAX_HEADER_VALUE_BYTE_LENGTH:
+            raise InvalidHeaderValueLength(len(utf8_string))
+
+        self._buffer.write(b"\07")
+        self._buffer.write(pack("!H", len(utf8_string)))
+        self._buffer.write(utf8_string)
+
+    def encode_timestamp(self, value: datetime.datetime):
+        timestamp_millis = int(value.timestamp() * 1000)
+        self._buffer.write(b"\x08")
+        self._buffer.write(pack("!q", timestamp_millis))
+
+    def encode_uuid(self, value: uuid.UUID):
+        self._buffer.write(b"\x09")
+        self._buffer.write(value.bytes)
+
+
+type _BytesLike = bytes | memoryview
+
+
+type _ArraySize = Literal[1] | Literal[2] | Literal[4]
+
+
+class DecodeUtils:
+    """Unpacking utility functions used in the decoder.
+
+    All methods on this class take raw bytes and return  a tuple containing the value
+    parsed from the bytes and the number of bytes consumed to parse that value.
+    """
+
+    UINT8_BYTE_FORMAT = "!B"
+    UINT16_BYTE_FORMAT = "!H"
+    UINT32_BYTE_FORMAT = "!I"
+    INT8_BYTE_FORMAT = "!b"
+    INT16_BYTE_FORMAT = "!h"
+    INT32_BYTE_FORMAT = "!i"
+    INT64_BYTE_FORMAT = "!q"
+    PRELUDE_BYTE_FORMAT = "!III"
+
+    # uint byte size to unpack format
+    UINT_BYTE_FORMAT: dict[_ArraySize, str] = {
+        1: UINT8_BYTE_FORMAT,
+        2: UINT16_BYTE_FORMAT,
+        4: UINT32_BYTE_FORMAT,
+    }
+
+    @staticmethod
+    def unpack_true(data: _BytesLike) -> tuple[bool, int]:
+        """This method consumes none of the provided bytes and returns True.
+
+        :param data: The bytes to parse from. This is ignored in this method.
+        :returns: The tuple (True, 0)
+        """
+        return True, 0
```

Comment:
This paradigm feels a little odd. We're throwing away the `data` bytes and assuming whatever was passed is `True`? Is the assumption this is never called unless we've already pre-validated the data? If that's the case, is there a reason to take the bytes as a parameter in the first place?

Reply:
This is a bit confusing I'll admit. It confused me too, and I forgot to add a comment to address it. I'll do that.

Basically, there are no bytes to read. Before any header is a type signifier, which of course determines how many bytes you need to read to get the value. There isn't a `bool` type signifier though, instead there's a `true` type signifier and a `false` type signifier, neither of which are followed by any data bytes. It's a space-saving measure.

Reply:
Actually I just removed these functions. They're terribly confusing and can be better inlined as lambdas in the header type map.